src/electionguard/precompute_buffers.cpp

```diff
@@ -183,122 +193,267 @@ namespace electionguard
                                                     quad->clone());
     }
 
-    void PrecomputeBufferContext::init(uint32_t size_of_queue /* = 0 */)
+#pragma endregion
+
+#pragma region PrecomputeBuffer
+
+    // Lifecycle Methods
+
+    PrecomputeBuffer::PrecomputeBuffer(const ElementModP &publicKey, uint32_t maxQueueSize,
+                                       bool shouldAutoPopulate)
+        : maxQueueSize(maxQueueSize == 0 ? DEFAULT_PRECOMPUTE_SIZE : maxQueueSize),
+          shouldAutoPopulate(shouldAutoPopulate), publicKey(publicKey.clone())
     {
-        std::lock_guard<std::mutex> lock(queue_lock);
-        getInstance().populate_OK = true;
+    }
+    PrecomputeBuffer::~PrecomputeBuffer() = default;
+
+    void PrecomputeBuffer::clear()
+    {
+        stop();
+
+        std::lock_guard<std::mutex> lock1(triple_queue_lock);
```

Comment:
there are now two lock guards, one for each collection. this actually sped up the unit tests a little bit. in the future we should re-enable async multithreading